src/net_processing.cpp

```diff
@@ -1965,6 +1969,107 @@ void static ProcessOrphanTx(CConnman* connman, CTxMemPool& mempool, std::set<uin
     }
 }
 
+/**
+ * Validation logic for compact filters request handling.
+ *
+ * May disconnect from the peer in the case of a bad request.
+ *
+ * @param[in]   pfrom           The peer that we received the request from
+ * @param[in]   chain_params    Chain parameters
+ * @param[in]   filter_type     The filter type the request is for. Must be basic filters.
+ * @param[in]   stop_hash       The stop_hash for the request
+ * @param[out]  stop_index      The CBlockIndex for the stop_hash block, if the request can be serviced.
+ * @param[out]  filter_index    The filter index, if the request can be serviced.
+ * @return                      True if the request can be serviced.
+ */
+static bool PrepareBlockFilterRequest(CNode* pfrom, const CChainParams& chain_params,
+                                      BlockFilterType filter_type,
+                                      const uint256& stop_hash,
+                                      const CBlockIndex*& stop_index,
+                                      BlockFilterIndex*& filter_index)
+{
+    const bool supported_filter_type =
+        (filter_type == BlockFilterType::BASIC &&
+         gArgs.GetBoolArg("-peercfilters", DEFAULT_PEERCFILTERS));
+    if (!supported_filter_type) {
+        LogPrint(BCLog::NET, "peer %d requested unknown block filter type: %d\n",
+                 pfrom->GetId(), static_cast<uint8_t>(filter_type));
+        pfrom->fDisconnect = true;
+        return false;
+    }
+
+    {
+        LOCK(cs_main);
+        stop_index = LookupBlockIndex(stop_hash);
+
+        // Check that the stop block exists and was at some point connected to the active chain.
+        if (!stop_index || !BlockRequestAllowed(stop_index, chain_params.GetConsensus())) {
```

Comment:
Is the following scenario plausible ? Client receives from peers block H1 as tip and goes offline. H1 is reorg'ed out of the active chain. After a month, Client goes back and try to start filter-sync through checkpoints. Sending H1 will be rejected as per `BlockRequestAllowed` we reject reorg'ed out block older than a month, but has been part of the active chain and dutifully announced.

It maybe a marginal issue, but at least comment should mention that stop block should have been connected to the active chain in the last month.

Reply:
Yes, that's possible. A client should always request the filter headers from a recently sync'ed block headers chain.

I was going to update the comment to say something like "Check that the stop block exists and is either in the best chain or was re-orged out less than a month ago", but I don't like repeating program logic in a comment, especially if it's located far from the code, since any changes in the logic would make the comment obsolete. Instead I've changed it to "Check that the stop block exists and the peer would be allowed to fetch it." Anyone reading the code can look up BlockRequestAllowed() for the exact conditions.